### ISSUE TITLE
Replace GameObject.SendMessage("Method");

### ISSUE THREAD

**Issue by HodgsonSDAS** [Microsoft/HoloToolkit-Unity#224]
Pull Request https://github.com/Microsoft/HoloToolkit-Unity/pull/244

Many[ experienced Unity developers](http://jacksondunstan.com/articles/3335) would probably tell you not use [GameObject.SendMessage("String");](https://docs.unity3d.com/ScriptReference/GameObject.SendMessage.html) because [it's slow](http://jacksondunstan.com/articles/3605) (under the hood it uses [reflection](http://jacksondunstan.com/articles/2972)), hard to debug, and it lacks direct symbol references in the editor which makes it prone to spelling errors _because_ it's a string and not a symbol. This can also make it difficult to find everywhere the method referenced can be called and could potentially increase the amount of spaghetti code.  It may be worth while to see this updated to utilize a [delegate/Event call instead](https://unity3d.com/learn/tutorials/topics/scripting/events?playlist=17117) or implement an Interface.

[Another good link that explains the difficulty with using SendMessage](http://www.sebaslab.com/whats-wrong-with-sendmessage-and-broadcastmessage-and-what-to-do-about-it/)

> Emblematic example of these bad choices is the communication system built around SendMessage and BroadcastMessage. If you use them, you should just stop already!
> 
> In fact SendMessage and BroadcastMessage are so wrong mostly because they heavily rely on reflection to find the functions to call. This of course has a given impact on performance, but performance is not the issue here, **_the issue is instead related to the quality of the resulting code_**.
> 
> What can be so bad about it? First (and foremost) the use of a string to recognize a function is way worse than using a string to recognize an Event. Just think about it: what happens if someone in the team decides to refactor the code where the calling function is defined and decides to rename the function or delete it?
> 
> I tell you what happens: the compiler will not warn the poor programmer of the mistake is doing and even worse, the code will just run as nothing happened. When the error will be found, it could be already too late.
> 
> Even worse is the fact that the calling function can be declared as private since the system uses reflection. You know what I usually do when I find a private function that is not used inside the declaring class? I just delete it, because the code is telling me: this function cannot be used outside this class and I am not using it, so why keep useless code?

## TL;DR
This will make the code easier to follow in the editor, easier to debug, and will increase performance.

## [Some arguments for using delegate events vs SendMessage](http://answers.unity3d.com/questions/185980/sendbroadcast-message-vs-delegates-and-events.html):

> SendMessage and delegates are used for similar purposes, but they work entirely differently. Delegates/Events behave exactly like they do in any other .Net program. SendMessage() is called much like Unity calls other messages like Update(). Unity reflects over all the code on the game object to see if it can invoke any messages.
> 
> As far as performance. Delegates are several times faster than SendMessage. In the order of performance:
> 
`(Least) Direct Method Call ==> Delegate/Event ==> Update() ==> SendMessage (Most Expensive)`
> As far as usage, SendMessage is like a very broad direct method call. The general paradigm is that one script specifically wants to send a message to a specific object. We're just not sure how many times is can respond or if it can respond at all. We know the intended receiver, we just don't know if it can receive the message.
> 
> With events, I usually follow the idea that the event owner doesn't know who or how many objects might respond. We are simply telling the world that something happened without too much knowledge of who will use this information.

Another major benefit of using Delegate/Event is type checking. A delegate specifies specific types that must be passed as parameters, while SendMessage just accepts an object.

## [A good argument for using Interfaces vs delegate events](http://jacksondunstan.com/articles/3043):

> Now that we’ve established which is faster—interfaces—the question becomes: should you bother? Interfaces are definitely more work to set up, harder to read, more error-prone, and more difficult to integrate into your code. The percentages faster are impressive, but will you save a lot of total CPU time? The above test results are measured in milliseconds per 100 million callback calls. That’s clearly way more than you’ll have in any reasonable game or app.

Although, one drawback to the Interface Implementation is that the method for the Interface will also be public, so anyone outside our class could potentially call OnTap (or whatever else we're implementing).

**Comment by jwittner** [comment]
I think there are multiple questions here:

1. Should we always use event/delegate going forward? _Probably._
1. Should we go back and add event/delegate functionality to SendMessage systems? _Would be nice!_
1. Should we drop support for existing SendMessage systems? _Meh._

**Comment by HodgsonSDAS** [comment]
> Should we drop support for existing SendMessage systems? Meh.

I can see this being the only real sticking point for some people.

**Comment by izatar** [comment]
Notice how OP could explain the send message with the code GameObject.SendMessage("Method");
but had to link to a video to explain how delegates work.

Making code more complicated and thus less accessable is not worth it for a few nanoseconds for an event that uses tens of thousands of times more for the recognizers.

This is a clear example of premature optimization.

**Comment by HodgsonSDAS** [comment]
@izatar did you have anything [constructive to add to your opinion](https://opensource.microsoft.com/codeofconduct/)?
Maybe references or code snippets you could contribute to further our discussion?

> This would be a clear example of premature optimization.

Yes, premature optimization can be bad, but we're talking about a [85.5x speed improvement](http://answers.unity3d.com/questions/327888/is-sendmessage-slow-.html).  Remember, the HoloLens is a mobile device, with a limited amount of overhead.  It's safe to say this should have been thought about during preproduction.

> We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.
- [Donald Knuth](http://c2.com/cgi/wiki?PrematureOptimization)

Debugging in VS with SendMessage is nearly impossible unless you know _exactly_ where to look for your trouble spots and place a break.  What what about those times where you're unsure, and stepping through the code is your only option?

Not to mention, while working at Microsoft last spring on a HoloLens project, we made the decision to _explicitly_ not use this design pattern and ensure we followed rules for exactly how we should be allocating memory.  In fact I'm surprised to see that many of those same rules are not enforced here.

Why don't you take the time to make the pull, and read the code, and write a constructive review.

**Comment by izatar** [comment]
My reference to "the OP" was not intended as a personal attack.
May I suggest that you could spend as much effort finding the weaknesses in your own argument as you did in mine?

**Comment by HodgsonSDAS** [comment]
I didn't think it was a personal attack by any means. I welcome criticism and critique. That's the only way for professionals like us to learn, grow, and hone our abilities.  Sorry, I've spent waaaaay too much time on the debate team.

**Comment by stbertou** [comment]
Hi there, 

First thanks and kudos for doing this "massive" work.. I've been thinking about doing something similar but never found the time.
I'll go through your PR asap but the only concern I have is on backwards compatibility issue.
Even though this design is better on many levels, right now getting the latest version of the HTK means ppl would have to update their code and that's IMO is a no go (your update is a major change and the fact there's no versioning system or release notes in place here doesn't help of course..)
But basically ppl would get the latest version and all the HL input wouldn't work anymore.
Also there's a lot of people who don't really care about performance, and arguably the SendMessage approach is ok for "simple" cases.
So my recommendation would be to keep the SendMessage API (and set it on per default) and add a new mode on the GazeManager etc.. (that was my plan but you were faster than me :-)
Any thoughts ?

**Comment by HodgsonSDAS** [comment]
@stbertou I totally agree, and even suggested keeping the old SendMessage stuff around, and marking it as obsolete to give people time to transition over.

Yes, I found that a lot of the time I was just rewriting the what what already in the HTK, but was frustrated by this because I wanted to stay open to future updates. That was my main driving factor behind these changes.

**Comment by NeerajW** [comment]
Thank you everyone for the hard work. 
We all should be respectful and open to each other's thoughts and comments. As said earlier, we all are learning and growing together.

Many people look at the HoloToolkit as a learning or starting point as well for Unity and HoloLens interaction. SendMessage is not ideal for reasons stated above but many developers find it super simple to understand and easy to use. Hence it exists and the reason it was used to teach concepts.

I am totally onboard for a transition plan. Such large breaking changes obviously won't work.
We need a better story for breaking changes and I will try to come up with a plan for that. 

Thanks again! Will review the PR shortly and sorry for the delay.

**Comment by HodgsonSDAS** [comment]
A couple of updates from the PR over the weekend:

KeywordManager will be removed from the PR while we look for other ways to facilitate assigning events in the Inspector.

GestureManager now throws a warning about deprecated SendMessage events and in the future will point to documentation supporting the best way to update.

**Comment by HodgsonSDAS** [comment]
Another major benefit of using Delegate/Event is type checking. A delegate specifies specific types that must be passed as parameters, while SendMessage just accepts an object.

**Comment by HodgsonSDAS** [comment]
[Another good link](http://www.sebaslab.com/whats-wrong-with-sendmessage-and-broadcastmessage-and-what-to-do-about-it/)

> Emblematic example of these bad choices is the communication system built around SendMessage and BroadcastMessage. If you use them, you should just stop already!
> 
> In fact SendMessage and BroadcastMessage are so wrong mostly because they heavily rely on reflection to find the functions to call. This of course has a given impact on performance, but performance is not the issue here, the issue is instead related to the quality of the resulting code.
> 
> What can be so bad about it? First (and foremost) the use of a string to recognize a function is way worse than using a string to recognize an Event. Just think about it: what happens if someone in the team decides to refactor the code where the calling function is defined and decides to rename the function or delete it?
> 
> I tell you what happens: the compiler will not warn the poor programmer of the mistake is doing and even worse, the code will just run as nothing happened. When the error will be found, it could be already too late.
> 
> Even worse is the fact that the calling function can be declared as private since the system uses reflection. You know what I usually do when I find a private function that is not used inside the declaring class? I just delete it, because the code is telling me: this function cannot be used outside this class and I am not using it, so why keep useless code?

**Comment by NeerajW** [comment]
**Path forward for this feature:**
1. This change is a breaking change for several apps. 
2. We have not collectively agreed on a the right design for the change. Delegates Vs Interface. 

Here's what I recommend we do for such features as documented in Contributing.md:

1. We will create a new branch called Feature-SendMessageRemoval.
2. @HodgsonSDAS please feel free to code this change in that branch. Only changes related to this feature will be accepted in that branch. Let's try to lock on the design as well.
3. We will give developers 2 weeks to move to the new feature once the PR's are accepted in that branch.
4. We will announce on forums and other sources where we can.
5. Then we will merge this breaking change after 2 weeks into master.
6. Feature-SendMessageRemoval branch will then be deleted.

Questions/concerns, please let me know.

**Comment by NeerajW** [comment]
@HodgsonSDAS a new branch called Feature-SendMessageRemoval has been created. Please feel free to use it for adding this feature. Also please close without merge the existing PR when ready. Thanks everyone!

**Comment by HodgsonSDAS** [comment]
Will do.

**Comment by HodgsonSDAS** [comment]
@darax I'd like to see how we could implement the Interface you were talking about instead of using the virtual methods for our Interactable class.  I see a lot of potential in that.

**Comment by HodgsonSDAS** [comment]
http://jacksondunstan.com/articles/3605

**Comment by keveleigh** [comment]
I don't disagree with this change, but that article doesn't really support removing SendMessage for performance reasons the way we use it. About SendMessage: 

> That said, each function type was called 100 million times, so it’s not so slow that you should never use it. Each call was only 6/10,000 of a millisecond, so feel free to make a few of these calls without worrying that it’ll ruin your app’s performance.

I'm pretty sure we call (most of? all of?) these one frame at a time. If we had a couple hundred every Update, maybe it'd be noticeable.

**Comment by HodgsonSDAS** [comment]
@keveleigh I think it's more about readability and debug ability more than anything, I just wanted to share that's all.

**Comment by keveleigh** [comment]
@HodgsonSDAS Ahh yep, definitely agreed with you there! It is interesting to see the data as well. Thanks for the share!

**Comment by HodgsonSDAS** [comment]
Closed by https://github.com/Microsoft/HoloToolkit-Unity/issues/277
